RFCs/0035-partiql-window-function.md

```diff
@@ -0,0 +1,1204 @@
+Next, we demonstrate how to perform operations over the produced binding tuple.
+Finally, we show that PartiQL’s window function semantic is backward compatible to SQL.
+
+## Motivation
+
+Window functions, also known as analytic OLAP functions, have been introduced as part of the SQL standard since 2003. Window functions allow users to express many useful query, such as time series analysis, ranking, moving averages, etc. Formulating such queries in plain SQL-92 can be cumbersome and inefficient, if even possible.
+
+This RFC purposes the semantics for **window partition** in PartiQL. Subsequent RFCs are expected to cover other concepts in window function such as **peer** and **window frame**.
+
+## Guide-level explanation
+
+### Out of scope
+
+* This RFC focus on explaining the semantics of PartiQL’s window functions. The PartiQL core syntax provided in this RFC is for guidance only. Finalizing grammar and syntax is out of the scope, excluding the supported SQL compatible window function. 
+* SQL’s window implicitly defines three concepts, partition, frame, and peer. As a first step, this RFC focus on partition only, the concept of frame and peer are out of scope for now, but some preliminary work has been done to make sure the other concepts can fit into the framework in a modular fashion.
+* An exhaustive mapping between SQL’s window function and PartiQL’s expression is out of the scope. In this RFC, we only define the mapping between SQL’s `lag` function and it's corresponding PartiQL expression as an example. The mapping between other functions are skipped.
+
+We assume that the readers of this RFC has basic understanding of SQL’s window function and PartiQL semantics.
+
+### Terminology
+
+1. RFC key word: MUST, MUST NOT, REQUIRED, SHALL, SHALL NOT, etc. See Appendix (1) for detail. 
+2. Windowed Table and Window: Windowed table is a defined in SQL spec 2011, section 4.15.14. A **windowed table** is a table together with one or more windows. A **window** is a transient data structure associated
+   with a table expression. A window is defined explicitly by a window definition or implicitly by an inline window specification. Implicitly defined windows have an implementation-dependent window name. A
+   **window** is used to specify window partitions and window frames, which are collections of rows used in the definition of **window functions**.
+3. Partition: To distinguish between SQL's definition of window, in this RFC, the term **partition** refers to the window partition specified for each row. See [SQL’s Windowed Table As Nested Data](#SQL’s-Windowed-Table-As-Nested-Data) for details. 
+4. Equivalent Partition: Two binding tuples $b, b' \in B^{in}_{WINDOWED}$ are in the same equivalence partition if and only if every partition expression $e_i$ evaluates to equivalent values $v_i$ (when evaluted on b) and $v_i'$ (when evaluated on $b'$). 
+
+### SQL’s Windowed Table As Nested Data
```

Comment:
I still see a disconnect between this section and the next; how can we make the connection to the next section more obvious? Does it make sense to have this section as a sub-section of the next section and explain its relation with the way we transform B<sub>in</sub><sup>WINDOWED</sup> to B<sub>out</sub><sup>WINDOWED</sup>?

Reply:
Thanks for the call out. I think it makes sense. Thinking about this more, the gap can come from the jump from SQL to partiQL to SQL again. Maybe we rename the section to `Partition As Nested Data`

```
### PartiQL’s Window Function Semantics

In the following section, we first define the semantics for `WINDOWED` clause in PartiQL.
The PartiQL `WINDOWED` clause can be thought of as a standalone operator that inputs a collection of binding tuples and outputs a collection of binding tuples.

Informally: 
Each binding tuple outputted by the `WINDOWED` clause contains "information" regarding: 1) the current binding tuple from the input, 2) the partition for the current binding tuple, 3) the index for access the current binding tuple in partition. 

This section proceed in four steps: 
 - Section [Partition As Nested Data] explains how to create a partition for each row/binding tuple, such partition created will be incorporated into the binding tuple produced by `WINDOWED` clause.  
 - Section [WINDOWED Clause](#WINDOWED-Clause) explains the core PartiQL WINDOWED structure and the binding tuples created by `WINDOWED` clause.
 - Section [Operations on Partition](#Operations-Over-Partition) explains how to perform operation over the produced partition data. 
 - Section [SQL Compatibility](#SQL-Compatibility) shows that SQL's window functions can be explained over by PartiQL's `WINDOWED` clause and expressions. 

#### Partition As Nested Data
In this section, we introduce a way to model partition as nested data. Notice that the nested data format, although not fitting in SQL's formal semantics, provides a simple and intuitive way to understand window function results, and forms the basis for PartiQL's window semantics. 
...
```

Reply:
+1 This is way better and reminds of the spec.'s language.

Reply:
Addressed in the new commit.